arrow/physical_layout/arrow.mojo

```diff
@@ -5,44 +5,88 @@ from arrow.buffer.bitmap import Bitmap
 from arrow.buffer.offset import OffsetBuffer64
 
 
-struct ArrowFixedWidthVector[T: AnyTrivialRegType]:
-    # TODO: support null values
+struct ArrowFixedWidthVector[T: DType]:
     var length: Int
     var null_count: Int
     var validity: Bitmap
-    var value: Pointer[UInt8]
-    var view: Pointer[T]
+    alias _ptr_type = DTypePointer[DType.uint8]
+    var value: Self._ptr_type
+    var view: DTypePointer[T]
 
-    var mem_use: Int
+    var mem_used: Int
 
-    fn __init__(inout self, values: List[T]):
+    fn __init__(inout self, values: List[Scalar[T]]):
         var byte_width = sizeof[T]()
         var num_bytes = len(values) * byte_width
         var num_bytes_with_padding = get_num_bytes_with_padding(num_bytes)
-        var ui8_ptr = Pointer[UInt8].alloc(
+        var ui8_ptr = Self._ptr_type.alloc(
             num_bytes_with_padding, alignment=ALIGNMENT
         )
         memset_zero(ui8_ptr, num_bytes_with_padding)
         var ptr = ui8_ptr.bitcast[T]()
 
-        var validity_list = List[Bool](len(values))
+        var validity_list = List[Bool](capacity=len(values))
 
         for i in range(values.size):
             validity_list.append(True)
             var val = values[i]
-            ptr.store(i, val)
+            ptr[i] = val
 
         self.value = ui8_ptr
         self.validity = Bitmap(validity_list)
         self.null_count = 0
         self.view = ptr
         self.length = len(values)
-        self.mem_use = num_bytes_with_padding
+        self.mem_used = num_bytes_with_padding
+
+    fn __init__(inout self, values: List[Optional[Scalar[T]]]):
+        var byte_width = sizeof[T]()
+        var num_bytes = len(values) * byte_width
+        var num_bytes_with_padding = get_num_bytes_with_padding(num_bytes)
+        var ui8_ptr = Self._ptr_type.alloc(
+            num_bytes_with_padding, alignment=ALIGNMENT
+        )
+        memset_zero(ui8_ptr, num_bytes_with_padding)
+        var ptr = ui8_ptr.bitcast[T]()
+
+        var validity_list = List[Bool](capacity=len(values))
+
+        var null_count = 0
+
+        for i in range(values.size):
+            var val = values[i]
+            if val:
+                validity_list.append(True)
+                ptr[i] = val.value()
+            else:
+                null_count += 1
+                validity_list.append(False)
+                ptr[i] = 0
+
+        self.value = ui8_ptr
+        self.validity = Bitmap(validity_list)
+        self.null_count = null_count
+        self.view = ptr
+        self.length = len(values)
+        self.mem_used = num_bytes_with_padding
+
+    fn __getitem__(self, index: Int) -> Optional[Scalar[T]]:
+        if not (0 <= index < self.length):
+            return None
```

Comment:
The change to account for optional values is nice but this check should still throw an out of bounds error. Not throwing an error here means that it's possible to have silent errors or unexpected behavior e.g. assuming `arr[-1]` would return the last item. The user mistakenly assumes it to be a valid and null value but it should be out of bounds.

Reply:
> Not throwing an error here means that it's possible to have silent errors or unexpected behavior e.g. assuming arr[-1] would return the last item.

that is what the Optional type is for, it has to be checked with bool(opt) before calling `.value()` on it

> The user mistakenly assumes it to be a valid and null value but it should be out of bounds.

they can't since it also checks the validity bitmap and any index could return None since it supports null values now.

Reply:
I can see where you are coming from, however at this point in the codebase we throw errors for out of bounds access instead of silent failures. To make the design change of returning values instead of throwing errors another separate pull request could be made. This pull request will only be merged if an error is thrown on out of bounds access